engine/atf.cpp

```diff
@@ -203,14 +203,12 @@ engine::atf_interface::exec_test(const model::test_program& test_program,
 /// \param test_program The test program to execute.
 /// \param test_case_name Name of the test case to invoke.
 /// \param vars User-provided variables to pass to the test program.
-/// \param unused_control_directory Directory where the interface may place
-///     control files.
 void
 engine::atf_interface::exec_cleanup(
     const model::test_program& test_program,
     const std::string& test_case_name,
     const config::properties_map& vars,
-    const fs::path& UTILS_UNUSED_PARAM(control_directory)) const
+    const fs::path& /* control_directory */) const
```

Comment:
I didn't know this was possible (leaving parameters unnamed when not used), but I'm not sure this is good. By removing the name, it's not possible to name this parameter any longer in the description—but the parameter exists.

Why didn't the current approach work?

Reply:
> Why didn't the current approach work?

First off, I felt a bit dirty adding in `unused_` prefixing with all UTILS_UNUSED_PARAM parameters, since it was separated from the actual macro implementation in `utils/defs.hpp`. I could readd it, but it seems a bit kludgy (especially when the upstream interface should be defining this).

Second off, omitting the name (specifying an "unnamed parameter") is a C++ compatible idiom that works with pre-C++11 (I used both clang++ and g++ to illustrate this):

```
cripps:kyua ngie$ (cd ~/nfs/git/scratch/programming/c++/; export CXX=g++; make clean; make)
rm -f map_demo map_demo regex_demo set_demo string_trim_demo unnamed_parameter_demo
...
g++ -g -W -Wall -Wunused -Werror -std=c++11 -std=c++98    unnamed_parameter_demo.cc   -o unnamed_parameter_demo
cripps:kyua ngie$ (cd ~/nfs/git/scratch/programming/c++/; ./unnamed_parameter_demo)
10
```

1. https://en.cppreference.com/w/cpp/language/function#Parameter_list
2. https://stackoverflow.com/a/12186734

Reply:
The rationale for `UTILS_UNUSED_PARAM` was two-fold: first, to silence the warning about unused parameters; but, second, to mangle the parameter name in a way that made it clear when it was put to use. Otherwise, one might forget the attribute attached to the parameter even when the parameter is later used.

I think your proposal achieves the same effect given that uncommenting the parameter name will trigger a warning if the parameter is actually unused.

Maybe a simpler approach would have been to make `UTILS_UNUSED_PARAM` expand to nothing? I just don't like seeing commented out "code" because nothing verifies it. But well, this is not a big deal here so I may just let it go your way.

Reply:
By the way, I'm guessing your change makes `UTILS_UNUSED_PARAM` unnecessary now? Would be nice to remove it altogether then.

Reply:
Moved to #198.